### DISCUSSION TITLE
Provide IdentityProviderMapper callback to handle IdP link removal

### DISCUSSION THREAD

**Discussion by thomasdarimont** [keycloak/keycloak#24762] [Ideas]
Currently, `IdentityProviderMapper` provides callback methods that are called when a new brokered identity is processed:
- org.keycloak.broker.provider.IdentityProviderMapper#preprocessFederatedIdentity
- org.keycloak.broker.provider.IdentityProviderMapper#importNewUser
- org.keycloak.broker.provider.IdentityProviderMapper#updateBrokeredUserLegacy
- org.keycloak.broker.provider.IdentityProviderMapper#updateBrokeredUser

Those methods handle the cases when a user is imported for the first time or when a known federated user signs in again. 
Some mappers manipulate the user in diverse ways, e.g., adding an attribute, assigning a role, or joining a group.

Problem:
Keycloak also supports removing an IDP federation link from a user. Currently, when an IDP link is removed, the user stays untouched, and there is no easy way to provide IDP mapper-specific clean-up logic to "undo" the user changes performed when the user logged in.

Current workaround:
Add custom logic to an event listener that handles the EventType.REMOVE_FEDERATED_IDENTITY, but this only addresses the case where a user removed an IDP link. When an admin removes the IDP link, the event is not triggered.

Solution:
I propose adding a new (default) method to the IdentityProviderMapper interface that developers can override in IdP mapper implementations to allow reverting the user change performed when the mapper during the initial execution.

The new method could look like this:

```java
    /**
     * Called after an Identity provider link is removed from the given user.
     * This allows some mapper specific cleanup code to be executed for the given user.
     *
     * @param session
     * @param realm
     * @param mapper
     * @param user
     */
    default void unlinkFederatedIdentity(KeycloakSession session, RealmModel realm, IdentityProviderMapperModel mapper, UserModel user) {
    }
```

With helper method like this, e.g. on KeycloakModelUtils, we could remove the federated identity provider link and run the "clean-up" logic from the unlink mappers afterwards.

```java
    public static boolean unlinkFederatedUser(KeycloakSession session, RealmModel realm, UserModel user, String idpAlias) {

        if (session.users().removeFederatedIdentity(realm, user, idpAlias)) {
            realm.getIdentityProviderMappersByAliasStream(idpAlias)
                    .forEach(mapperModel -> {
                        IdentityProviderMapper mapper = (IdentityProviderMapper) session.getKeycloakSessionFactory().getProviderFactory(IdentityProviderMapper.class, mapperModel.getIdentityProviderMapper());
                        mapper.unlinkFederatedIdentity(session, realm, mapperModel, user);
                    });
            return true;
        }

        return false;
    }
```

The method `KeycloakModelUtils.unlinkFederatedUser(session, realm, user, providerAlias);`would then need to be called from two places:
- org.keycloak.services.resources.account.LinkedAccountsResource#removeLinkedAccount (called when a user removes a federated identity link)
- org.keycloak.services.resources.admin.UserResource#removeFederatedIdentity (called when an admin removes a federated identity link)

This could be used to do cleanup operations like the following:
- Remove an attribute
- Remove a role from a user
- Remove groups from a user
- Send a password reset link to a user
- Add a new required action for the user

**Comment by danielFesenmeyer** [comment]
Sounds reasonable :thumbsup:

Some further thoughts on this based on my understanding of IDPs and mappers:

When adding the method `unlinkFederatedIdentity` to the mapper, it would be great when it's also implemented on the existing mappers. Currently we would have to subclass existing mappers to get that behavior.

It could also help when the "unlink-behavior" could be toggled at least on the mapper, maybe even on the IDP - similar to the "sync-mode" behavior.
This toggle could help when a user has multiple IDPs linked, which use the same mapper with same config. In this case, when IDP1 gets unlinked, a mapper configured for IDP1 could remove data configured for IDP2.
For cleanup in such case a custom "IDP1-Unlink-Mapper" could be implemented, that just provides the unlink-operation, which only removes the attributes relevant for IDP1.
